core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/EventContextServices.java

```diff
@@ -0,0 +1,414 @@
+import org.polarsys.capella.core.data.interaction.InstanceRole;
+import org.polarsys.capella.core.data.interaction.InteractionFragment;
+import org.polarsys.capella.core.data.interaction.InteractionState;
+import org.polarsys.capella.core.data.interaction.MessageEnd;
+import org.polarsys.capella.core.data.interaction.Scenario;
+import org.polarsys.capella.core.data.interaction.SequenceMessage;
+import org.polarsys.capella.core.data.interaction.StateFragment;
+import org.polarsys.capella.core.data.interaction.TimeLapse;
+
+/**
+ * Compute EventContext structure for sequence diagram.
+ * 
+ * @author nlepine
+ *
+ */
+public class EventContextServices {
+
+  private EventContextServices() {
+    super();
+  }
+
+  /**
+   * Compute EventContext structure from InstanceRole if needed and put it in cache.
+   * 
+   * @param instanceRole
+   *          InstanceRole
+   * @return List<EventContext>
+   */
+  public static List<EventContext> getEventContexts(InstanceRole instanceRole) {
+    List<EventContext> eventContexts = getInstanceRoleToEventContextCache(instanceRole);
+    if (eventContexts == null) {
+      eventContexts = computeInstanceRoleEventContextStructure(instanceRole);
+      if (isRefreshCacheEnabled()) {
+        putInstanceRoleToEventContextsCache(instanceRole, eventContexts);
+      }
+    }
+    return eventContexts;
+  }
+
+  /**
+   * Get direct event on element. (InstanceRole -> Executions + StateFragment, Execution -> Execution children +
+   * StateFragment)
+   * 
+   * @param element
+   *          EObject
+   * @param instanceRole
+   *          InstanceRole
+   * @return Collection<EObject>
+   */
+  public static Collection<EObject> getDirectEvents(EObject element, InstanceRole instanceRole) {
+    Collection<EObject> directEvents = getDirectEventsFromCache(element);
+    if (directEvents == null) {
+      // compute result from EventContext structure
+      List<EventContext> eventContexts = getEventContexts(instanceRole);
+
+      directEvents = eventContexts.stream().filter(eventContext -> element.equals(eventContext.getParent()))
+          .map(EventContext::getElement).filter(event -> event != element).distinct().collect(Collectors.toList());
+      if (isRefreshCacheEnabled()) {
+        putDirectEventsInCache(element, directEvents);
+      }
+    }
+    return directEvents;
+  }
+
+  /**
+   * Get Execution direct events on element.
+   * 
+   * @param element
+   *          EObject
+   * @param instanceRole
+   *          InstanceRole
+   * @return Collection<EObject>
+   */
+  public static Collection<Execution> getExecutionDirectEvents(EObject element, InstanceRole instanceRole) {
+    // get executions from cache
+    Collection<EObject> directEvents = getDirectEventsFromCache(element);
+    if (directEvents != null) {
+      return directEvents.stream().filter(Execution.class::isInstance).map(Execution.class::cast)
+          .collect(Collectors.toList());
+    }
```

Comment:
This code block  does not serve any purpose, since the `getDirectEvents()` at line 122 already does verify if the information is available in the cache. This block should thus be removed.

Reply:
this code filters the directEvents and return only the executions.